posts/2018-07-12-servant-dsl-typelevel.md

```diff
@@ -0,0 +1,664 @@
+but to keep track of them at the type level too, even with this design. Because
+we need to make sure suitable encoding/decoding instances are available for the
+types that will be represented with those MIME types, and this cannot be done when
+discovering `"application/json"` in a list somewhere, at runtime.
+
+All in all, there is no value in keeping anything at the value level at this point.
+And we are already traversing a bunch of types mixed together with funny symbols and
+computing the type of a link making function as we go,
+as evidenced by the `HasLink` instances from above, so we've already got one
+foot in type-level land.
+
+An important tradeoff that we are making here is that while putting more
+information at the type-level indeed makes things more complex, it does
+however give a chance to our descriptions to influence more things, including
+other types. This is noticeable in the last `HasLink` instance we wrote, where
+making `Capture` track the type the url fragment is going to be decoded to
+allowed us to directly make the link-making function take a value of that type,
+instead of a string. This is strictly more powerful and will allow us to work
+in a very strongly typed world and where the typechecker "writes" a whole lot of
+code for us.
+
+Let's bite the bullet and finally take a quick look at what servant's type-level
+approach looks like.
+
+# Servant's approach (simplified)
+
+First, let me emphasize that any of the designs we have considered so far
+are interesting on their own and are fruitful in different ways. They simply
+were not quite good enough to meet our requirements which were, again, dictated
+by the projects and needs we had at work. This whole project started because
+we were sick of getting things wrong when manually constructing (client) or
+deconstructing (server) HTTP requests and so on.
+
+Now, let's write our type-level DSL. If you want a longer version of just this
+section, with more explanations, you may want to read
+[Implementing a minimal version of servant](https://www.well-typed.com/blog/2015/11/implementing-a-minimal-version-of-haskell-servant/).
+
+``` haskell
+-- GHC-flavoured Haskell supports type-level strings, they are the only type-level
+-- entity to have kind Symbol. We will therefore just use those,
+-- wrapped with 'Static'. See the first link in the "Going further"
+-- section if you're not familiar with type-level strings, kinds, etc.
+data Static (str :: Symbol)
+data Capture a
+
+-- GHC-flavoured Haskell (with the DataKinds extension) supports promoting ordinary data types
+-- to kinds and their constructors to types of those kinds. See again the
+-- first link in the "Going further" section if this is new to you.
+-- In our example, this lets us parametrize Verb not by an ordinary type but by
+-- a constructor of 'Method'. Indeed, 'method' can only be instantiated to
+-- Get and Post.
+data Method = Get | Post
+data Verb (method :: Method)
+
+infixr 5 :>
+data a :> b
```

Comment:
I'd specify kinds here too. `(a :: k) :> (b :: Type)` with an explanation that we need polykinded type on the left for symbols, but bare `Type` is enough on the left. As a side note, that restriction helps more often than you think

Reply:
actually here it's `Type -> Type -> Type`, isn't it?

Reply:
Yes, I purposefully do static strings with a `*`-kinded wrapper to avoid the need for poly-kinds :-)